From d453c08a40c7c72135eb10ca73dc365840ef430d Mon Sep 17 00:00:00 2001
From: Igor Grinberg <igrinberg@habana.ai>
Date: Thu, 06 Jun 2019 16:13:06 +0300
Subject: [PATCH] [SW-4185] gaudi: pll: support w/a w/o gating

The PCI PLL is configured partially during BTL time and includes gating
of HBW and LBW, and thus no (re)configuration is needed during the
Preboot time.
BTL cannot configure the PCI PLL DIV2 output as it will disable the
flash access while running from it and thus the PCI PLL DIV2 and DIV3
along, should be done in Preboot.
Add the new HL_PLL_WA_NOGT enum hl_pll_wa_sel value to specify the no
gated W/A for PLLs like the PCI.
Also rename the hl_pll_gated_clk_sw() function to hl_pll_wa_clk_sw(), so
it will better reflect its new functionality.

Change-Id: I64e05b3b8f7565646fd658fce7fe730ce65e0faa
Signed-off-by: Igor Grinberg <igrinberg@habana.ai>
---

diff --git a/common/pll.c b/common/pll.c
index 6e22c16..eb6ef92 100644
--- a/common/pll.c
+++ b/common/pll.c
@@ -107,6 +107,7 @@
 		reg = cfg->base + HL_PLL_DIV_SEL + (sel * sizeof(u32));
 		switch (cfg->div_wa[sel]) {
 		case HL_PLL_WA_INIT:
+		case HL_PLL_WA_NOGT:
 			continue;
 		case HL_PLL_WA_NOWA:
 		default:
diff --git a/common/pll.h b/common/pll.h
index 0610057..5f77209 100644
--- a/common/pll.h
+++ b/common/pll.h
@@ -43,6 +43,7 @@
 enum hl_pll_wa_sel {
 	HL_PLL_WA_NOWA = 0,
 	HL_PLL_WA_INIT,
+	HL_PLL_WA_NOGT,
 };
 
 enum hl_pll_div_out {
diff --git a/gaudi/asic.c b/gaudi/asic.c
index 2e254a0..77d17ce 100644
--- a/gaudi/asic.c
+++ b/gaudi/asic.c
@@ -352,30 +352,40 @@
 	return ~gmask & 0xF;
 }
 
-static void hl_pll_gated_clk_sw(struct hl_pll_cfg *cfg)
+static void hl_pll_wa_clk_sw(struct hl_pll_cfg *cfg)
 {
 	int i;
 	u32 gmask = hl_pll_build_gate_mask(cfg);
 	u64 reg;
 
-	if (gmask == 0xF) {
-		preboot_puts("Gate mask is 0xF - no gate, skipping...\n");
-		return;
+	if (gmask != 0xF) {
+		regmap_writel(gmask, cfg->base + HL_PLL_CLK_GATER);
+	} else {
+		preboot_puts_raw("PLL ");
+		preboot_puts_raw(cfg->name);
+		preboot_puts(" Gate mask is 0xF - skip gating...\n");
 	}
-
-	regmap_writel(gmask, cfg->base + HL_PLL_CLK_GATER);
 	hl_pll_udelay(1);
 
 	for (i = 0; i < HL_PLL_DIV_SIZE; i++) {
-		if (cfg->div_wa[i] == HL_PLL_WA_INIT) {
-			reg = cfg->base + HL_PLL_DIV_SEL + (i * sizeof(u32));
-			regmap_writel(cfg->div_sel[i], reg);
+		if (cfg->div_wa[i] == HL_PLL_WA_NOWA)
+			continue;
+
+		reg = cfg->base + HL_PLL_DIV_SEL + (i * sizeof(u32));
+		/* if DIV_PLL - first change DIV_REF, then DIV_PLL */
+		if (cfg->div_sel[i] == HL_PLL_DIV_PLL) {
+			regmap_writel(HL_PLL_DIV_REF, reg);
+			hl_pll_udelay(1);
 		}
+
+		regmap_writel(cfg->div_sel[i], reg);
 	}
 	hl_pll_udelay(10);
 
-	regmap_writel(0xF, cfg->base + HL_PLL_CLK_GATER);
-	hl_pll_udelay(10);
+	if (gmask != 0xF) {
+		regmap_writel(0xF, cfg->base + HL_PLL_CLK_GATER);
+		hl_pll_udelay(10);
+	}
 }
 
 void hl_pll_wa(void)
@@ -400,6 +410,6 @@
 			break;
 		}
 
-		hl_pll_gated_clk_sw(&gaudi_pll_cfg[i]);
+		hl_pll_wa_clk_sw(&gaudi_pll_cfg[i]);
 	}
 }
